### ISSUE TITLE
ICE: `Unimplemented` selecting `Binder(...)`

### ISSUE THREAD

**Issue by ptal** [rust-lang/rust#24294]
Since it's an ICE I think it must be reported even if the message `Unimplemented` is pretty clear. However I don't know what is the problem, the failing code is [here](https://github.com/ptal/Rust.mcp/commit/07191277811bb076842d4badf4e10c4bb4c31640) when compiling with `cargo test`.

I guess the problem is related to what I try to achieve (more on [rust.users](http://users.rust-lang.org/t/generic-programming-with-traits/908/1)) because it was working before I extract `deep_clone` into another trait (see propagator.rs).

The output is:

```
➜  Rust.mcp git:(master) ✗ RUST_BACKTRACE=1 ct
   Compiling rustc-serialize v0.3.12
   Compiling libc v0.1.6
   Compiling log v0.3.1
   Compiling rand v0.3.7
   Compiling num v0.1.22
   Compiling intervallum v0.2.0
   Compiling mcp v0.1.0 (file:///home/ptalbot/repositories/Rust.mcp)
src/libmcp/lib.rs:1:1: 1:1 error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(solver::propagator::Propagator<<solver::fd::var::FDVar<interval::interval::Interval<i32>> as solver::variable::Variable>::Event>)` during trans
src/libmcp/lib.rs:1 // Copyright 2015 Pierre Talbot (IRCAM)
                    ^
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/ptalbot/repositories/installed/rust/src/libsyntax/diagnostic.rs:131

stack backtrace:
   1:     0x7f84e4805aa9 - sys::backtrace::write::h94b87a491cd42a0dVQC
   2:     0x7f84e482a377 - panicking::on_panic::h3c97f761284d87b6ThJ
   3:     0x7f84e4773eb2 - rt::unwind::begin_unwind_inner::h50475333015d4c8e1WI
   4:     0x7f84e1d3279d - rt::unwind::begin_unwind::h13395023725706001170
   5:     0x7f84e1d32732 - diagnostic::SpanHandler::span_bug::h667822574712ce9ddqB
   6:     0x7f84e25d04d3 - session::Session::span_bug::hac32b6dd59313c6c4kr
   7:     0x7f84e3f2cd53 - trans::common::fulfill_obligation::h46adc7065de8f484gql
   8:     0x7f84e3fa5c22 - trans::meth::get_vtable::closure.47139
   9:     0x7f84e3fa711d - iter::FlatMap<I, U, F>.Iterator::next::h1778182984942007577
  10:     0x7f84e3fa6c33 - vec::Vec<T>.FromIterator<T>::from_iter::h6096893015971501869
  11:     0x7f84e3ef5014 - trans::meth::get_vtable::h8a559355328a84f42gz
  12:     0x7f84e3ef199f - trans::expr::unsized_info::h5694840028648445667
  13:     0x7f84e3efba08 - trans::expr::apply_adjustments::apply_autoref::ha5b014354888e0cbxGh
  14:     0x7f84e3eac460 - trans::expr::trans::hd8d26ce886b0564aanh
  15:     0x7f84e3eaa7be - trans::expr::trans_into::hf4e81c9478133f9ePgh
  16:     0x7f84e3eab858 - trans::controlflow::trans_block::h0730db252adca761g8d
  17:     0x7f84e3f66171 - trans::base::trans_closure::h86ff45a647f9a57aovt
  18:     0x7f84e3e9ab6a - trans::base::trans_fn::hc2ddacf6da17d53a7Ft
  19:     0x7f84e3e9b873 - trans::monomorphize::monomorphic_fn::h56d1ae720d85e49byvd
  20:     0x7f84e3ed240b - trans::callee::trans_fn_ref_with_substs::h01c687e4ee2c152aydg
  21:     0x7f84e3faa405 - trans::meth::emit_vtable_methods::closure.47200
  22:     0x7f84e3fa6104 - trans::meth::get_vtable::closure.47139
  23:     0x7f84e3fa711d - iter::FlatMap<I, U, F>.Iterator::next::h1778182984942007577
  24:     0x7f84e3fa6c91 - vec::Vec<T>.FromIterator<T>::from_iter::h6096893015971501869
  25:     0x7f84e3ef5014 - trans::meth::get_vtable::h8a559355328a84f42gz
  26:     0x7f84e3ef199f - trans::expr::unsized_info::h5694840028648445667
  27:     0x7f84e3efba08 - trans::expr::apply_adjustments::apply_autoref::ha5b014354888e0cbxGh
  28:     0x7f84e3eac460 - trans::expr::trans::hd8d26ce886b0564aanh
  29:     0x7f84e3eddb70 - trans::callee::trans_args::h251f356823655b63ITg
  30:     0x7f84e3ee4267 - trans::callee::trans_call_inner::h6725593984957005344
  31:     0x7f84e3eeaf26 - trans::expr::trans_rvalue_dps_unadjusted::h1c946d6402e88142Vui
  32:     0x7f84e3eaaa17 - trans::expr::trans_into::hf4e81c9478133f9ePgh
  33:     0x7f84e3eaa08f - trans::controlflow::trans_stmt_semi::hfe89ef1ae15dd80bk7d
  34:     0x7f84e3eab512 - trans::controlflow::trans_block::h0730db252adca761g8d
  35:     0x7f84e3f66171 - trans::base::trans_closure::h86ff45a647f9a57aovt
  36:     0x7f84e3e9ab6a - trans::base::trans_fn::hc2ddacf6da17d53a7Ft
  37:     0x7f84e3e96b16 - trans::base::trans_item::hfddb1fbf67b05a3fn4t
  38:     0x7f84e3e97258 - trans::base::trans_item::hfddb1fbf67b05a3fn4t
  39:     0x7f84e3e97258 - trans::base::trans_item::hfddb1fbf67b05a3fn4t
  40:     0x7f84e3e97258 - trans::base::trans_item::hfddb1fbf67b05a3fn4t
  41:     0x7f84e3f6ce18 - trans::base::trans_crate::h5c8bb29a766697e7T0u
  42:     0x7f84e4e85d2a - driver::phase_4_translate_to_llvm::h41e17ac30c52776ejOa
  43:     0x7f84e4e5db6a - driver::compile_input::h5f55fe906b0b0908Qba
  44:     0x7f84e4f1c391 - run_compiler::h9ad6d857f076047f14b
  45:     0x7f84e4f19fe2 - boxed::F.FnBox<A>::call_box::h10211653195867772721
  46:     0x7f84e4f19519 - rt::unwind::try::try_fn::h77990541018311353
  47:     0x7f84e48a8db8 - rust_try_inner
  48:     0x7f84e48a8da5 - rust_try
  49:     0x7f84e4f197c8 - boxed::F.FnBox<A>::call_box::h10407204623776454161
  50:     0x7f84e4817cf1 - sys::thread::create::thread_start::haa4289bda09d9217bSH
  51:     0x7f84df1c5373 - start_thread
  52:     0x7f84e43ed27c - __clone
  53:                0x0 - <unknown>

Could not compile `mcp`.

To learn more, run the command again with --verbose.
```

**Comment by ptal** [comment]
The commit https://github.com/ptal/Rust.mcp/commit/be05cf0cfea4a7922975478aaa8df9bb02602c40 make the ICE dissapeared. It certainly linked to the `Event` associated type of the trait `Variable` that I extracted outside of the trait in this commit.

**Comment by steveklabnik** [comment]
Triage: it's been a very long time, and the code no longer builds at all. Since nobody else has commented in over a year, I'm going to give this one a close. if you see this ICE reappear, please open a new issue, thanks!
